plutus-scb/src/Plutus/SCB/Core/ContractInstance.hs

```diff
@@ -0,0 +1,484 @@
+    , Member WalletEffect effs
+    )
+    => Eff effs ()
+processOwnPubkeyRequests = do
+    logInfo "processOwnPubkeyRequests start"
+    state <- Query.contractStates <$> runGlobalQuery (Query.ownPubKeyRequests @t)
+    let go pk (i, (it, ())) = do
+            logDebug . render $ "processOwnPubkeyRequest" <+> pretty i <+> pretty pk
+            sendContractInboxMessage @t (OwnPubkeyResponse pk) i it
+    unless (Map.null state) $ do
+        pk <- ownPubKey
+        traverse_ (go pk) (Map.toList state)
+    logInfo "processOwnPubkeyRequests end"
+
+processAwaitSlotRequests ::
+    forall t effs.
+    ( Member Log effs
+    , Member (EventLogEffect (ChainEvent t)) effs
+    , Member WalletEffect effs
+    )
+    => Eff effs ()
+processAwaitSlotRequests = do
+    logInfo "processAwaitSlotRequests start"
+    state <- Query.contractStates <$> runGlobalQuery (Query.awaitSlotRequests @t)
+    unless (Map.null state) $ do
+        slot <- walletSlot
+        traverse_ (uncurry $ processAwaitSlotRequest @t slot) (Map.toList state)
+    logInfo "processAwaitSlotRequests end"
+
+processAwaitSlotRequest ::
+    forall t effs.
+    ( Member (EventLogEffect (ChainEvent t)) effs
+    , Member Log effs
+    )
+    => Slot
+    -> ContractInstanceId
+    -> (ContractIteration, WaitingForSlot)
+    -> Eff effs [ChainEvent t]
+processAwaitSlotRequest currentSlot i (it, WaitingForSlot (Just targetSlot))
+    | currentSlot >= targetSlot = do
+        logDebug . render $ "processAwaitSlotRequest" <+> pretty i <+> pretty targetSlot
+        sendContractInboxMessage @t (AwaitSlotResponse currentSlot) i it
+processAwaitSlotRequest _ _ _ = pure []
+
+processUtxoAtRequests ::
+    forall t effs.
+    ( Member (EventLogEffect (ChainEvent t)) effs
+    , Member ChainIndexEffect effs
+    , Member Log effs
+    )
+    => Eff effs ()
+processUtxoAtRequests = do
+    logInfo "processUtxoAtRequests start"
+    state <- Query.contractStates <$> runGlobalQuery (Query.utxoAtRequests @t)
+    unless (Map.null state) $ do
+        utxos <- watchedAddresses
+        traverse_ (uncurry $ processUtxoAtRequest @t utxos) (Map.toList state)
+    logInfo "processUtxoAtRequests end"
+
+processUtxoAtRequest ::
+    forall t effs.
+    ( Member (EventLogEffect (ChainEvent t)) effs
+    , Member ChainIndexEffect effs
+    , Member Log effs
+    )
+    => AddressMap
+    -> ContractInstanceId
+    -> (ContractIteration, AddressSet)
+    -> Eff effs [ChainEvent t]
+processUtxoAtRequest utxos i (it, AddressSet addresses) =
+    let go address = do
+            logDebug . render $ "processUtxoAtRequest" <+> pretty i <+> pretty address
+            startWatching address
+            let response = UtxoAtAddress
+                    { address = address
+                    , utxo    = view (AM.fundsAt address) utxos
+                    }
+            sendContractInboxMessage @t (UtxoAtResponse response) i it
+    in fold <$> traverse go (Set.toList addresses)
```

Comment:
Seems like this should be a `foldMap` variant but I can't do it in my head...

Reply:
hoogle says there's `foldMapM` in various places but nowhere standard.

Reply:
yeah that always seemed like an omission...